### PR TITLE
Change asset cache key

### PR DESCRIPTION
**Pull request by typfel**
## What's new in this PR?

- Base cache keys on conversation and sender in addition to the nonce.
- Changed `ZMMessage` initializer to require a `nonce`
- The `nonce` on the `ZMMessage` is now marked as `nullable`.
- Made helper methods for creating messages internal in order to force dependencies to use the public API:s for adding messages to a conversation. This should hopefully make tests less prone to break in future.

## Dependencies

Needs releases with:

- [x] https://github.com/wireapp/wire-ios-utilities/pull/36

## Notes

Would have like to have `nonce` be `nonull`. Well investigate a bit more to see if it's safe.

## Todo
- [x] Decide on a caching solution.
- [x] Investigate `nonce` nullability.

### CODE REVIEW

File: Source/Model/Conversation/ZMConversation.m [hunk 1]
```diff
@@ -1491,7 +1499,7 @@ - (NSUInteger)sortedAppendMessage:(ZMMessage *)message;
         self.lastModifiedDate = message.serverTimestamp;
     }
     
-    [message prepareToSend];
+//    [message prepareToSend];
```
Comment (mikeger) [inline Source/Model/Conversation/ZMConversation.m]: Debug comment?

File: Source/Model/Message/AssetCache.swift [hunk 2]
```diff
@@ -21,20 +21,20 @@ import WireSystem
...
+//    public var zm_imageAssetCache : ImageAssetCache! {
+//        get {
+//            return self.userInfo[NSManagedObjectContextImageAssetCacheKey] as? ImageAssetCache
+//        }
+//        
+//        set {
+//            self.userInfo[NSManagedObjectContextImageAssetCacheKey] = newValue
+//        }
+//    }
+//}
```
Comment (mikeger) [inline Source/Model/Message/AssetCache.swift]: Most likely not necessary any more?

File: Source/Model/Message/AssetCache.swift [hunk 3]
```diff
@@ -108,52 +108,53 @@ struct PINAssetCache : Cache {
...
+//        let encryptedComponent = encrypted ? "_encrypted" : ""
+//        return "\(messageID.transportString())_\(String(describing: tagComponent))\(encryptedComponent)"
+//    }
+//}
+//
+//public extension ImageAssetCache {
+//    func wipeCache() {
+//        cache.wipeCache()
+//    }
+//}
```
Comment (mikeger) [inline Source/Model/Message/AssetCache.swift]: Most likely not necessary any more?

File: Source/Model/Message/AssetEncryption.swift [hunk 4]
```diff
@@ -81,58 +81,78 @@ extension Cache {
...
+//        return self.cache.decryptAssetIfItMatchesDigest(plaintextCacheKey, encryptedEntryKey: encryptedCacheKey, encryptionKey: encryptionKey, sha256Digest: sha256Digest)
+//    }
+//    
+//    /// Encrypts a plaintext cache entry to an encrypted one, also computing the digest of the encrypted entry
+//    public func encryptFileAndComputeSHA256Digest(_ nonce: UUID, format: ZMImageFormat) -> ZMImageAssetEncryptionKeys? {
+//        let plaintextCacheKey = type(of: self).cacheKeyForAsset(nonce, format: format, encrypted: false)
+//        let encryptedCacheKey = type(of: self).cacheKeyForAsset(nonce, format: format, encrypted: true)
+//        return self.cache.encryptFileAndComputeSHA256Digest(plaintextCacheKey, encryptedEntryKey: encryptedCacheKey)
+//    }
+//}
```
Comment (mikeger) [inline Source/Model/Message/AssetEncryption.swift]: Most likely not necessary any more?

File: Tests/Source/Model/Messages/AssetCacheTests.swift [hunk 5]
```diff
@@ -41,7 +41,7 @@ class FileAssetCacheTests: CacheTests {}
 
 
 // MARK: - Storing and retrieving image assets
-extension ImageAssetCacheTests {
+extension ImageAssetCacheTests { // TODO jacob removed from target
```
Comment (mikeger) [inline Tests/Source/Model/Messages/AssetCacheTests.swift]: TODO left

**Comment by mikeger** [comment]
Good ol' sync engine refactoring days!! 👍

**Comment by typfel** [comment]
So I think nonce could be `nonnull` since we were always setting a nonce in `willSave` so even local system message should have a nonce.

**Comment by typfel** [comment]
@vytis @mikeger remember why had this behaviour https://github.com/wireapp/wire-ios-data-model/pull/436/commits/97a87f48b57781b3327a19188f56688822086c66#diff-d9b8d4f5dd3f3e72ccafe339f45b2fa8R1530 ?

**Comment by vytis** [comment]
From my limited search it seems that it was used as cache key at some point (v2 assets), so makes sense that it would change after data is changed - 
https://github.com/wireapp/wire-ios-data-model/blob/0105feaeeb2ad322f57440b545f061c3c28d97c4/Source/Model/Message/V2Asset.swift#L64

https://github.com/wireapp/wire-ios-data-model/blob/8d584785a3c339819c9654ebdc1d9d23fe2f4602/Tests/Source/Model/Messages/ZMClientMessageTests%2BTextMessage.swift#L129

**Comment by typfel** [comment]
@vytis 
The cache key is only used for the in-memory cache in UI AFAIK. Changing the cache key would only be necessary if the image becomes invalid which is only the case when a user changes his/her profile image.

For normal messages the assets are not allowed to be replaced after they've been uploaded. That's why hoping I could get away with using the FileAssetCache cache keys also for the image identifiers.

File: Source/Model/Message/FileAssetCache.swift [hunk 6]
```diff
@@ -163,61 +163,118 @@ open class FileAssetCache : NSObject {
...
+    static func cacheKeyForAsset(_ message : ZMConversationMessage, identifier: String? = nil, encrypted: Bool = false) -> String? {
+        guard let messageId = message.nonce?.transportString(),
+              let senderId = message.sender?.remoteIdentifier?.transportString(),
+              let conversationId = message.conversation?.remoteIdentifier?.transportString()
         else {
-            return "\(messageID.transportString())_\(suffix)"
+            return nil
+        }
+    
+        var key = "\(messageId)_\(senderId)_\(conversationId)"
```
Comment (vytis) [inline Source/Model/Message/FileAssetCache.swift]: This maybe could be done slightly more simple by having array of strings and joining it with separator afterwards.

File: Source/Model/Message/FileAssetCache.swift [hunk 7]
```diff
@@ -163,61 +193,118 @@ open class FileAssetCache : NSObject {
...
+              let senderId = message.sender?.remoteIdentifier?.transportString(),
+              let conversationId = message.conversation?.remoteIdentifier?.transportString()
         else {
-            return "\(messageID.transportString())_\(suffix)"
+            return nil
         }
+    
+        var key = "\(messageId)_\(senderId)_\(conversationId)"
+        
+        if let identifier = identifier {
```
Comment (vytis) [inline Source/Model/Message/FileAssetCache.swift]: (comment moved from another place):
This maybe could be done slightly more simple by having array of optional strings, flatmapping it and joining with separator afterwards.

File: Source/Model/Message/ZMAssetClientMessage+FileMessageData.swift [hunk 8]
```diff
@@ -115,7 +115,24 @@ extension ZMAssetClientMessage: ZMFileMessageData {
...
+        }
+        
+        do {
+            try FileManager.default.createDirectory(at: temporaryFileURL.deletingLastPathComponent(), withIntermediateDirectories: true, attributes: nil)
+            try FileManager.default.linkItem(at: assetURL, to: temporaryFileURL)
+        } catch {
+            return nil
+        }
+        
+        return temporaryFileURL;
```
Comment (mikeger) [inline Source/Model/Message/ZMAssetClientMessage+FileMessageData.swift]: Semicolon 😰

File: Source/Model/Message/ZMMessage+Internal.h [hunk 9]
```diff
@@ -167,7 +169,7 @@ inManagedObjectContext:(NSManagedObjectContext * _Nonnull)moc;
 @interface ZMMessage ()
 
 @property (nonatomic) NSString * _Nullable senderClientID;
-@property (nonatomic) NSUUID * _Null_unspecified nonce;
+@property (nonatomic) NSUUID * _Nullable nonce;
```
[the remark this answers is not in the document]
Comment (typfel) [reply Source/Model/Message/ZMMessage+Internal.h]: I still think we can make it `nonull` but changing will affect so many tests that I leaning towards doing it in a separate PR.

**Pull request merged**